src/cpp/src/config.cpp

```diff
@@ -4,246 +4,392 @@
+                            if (section == v.s_section && option_name == v.s_option) {
+                                *(v.var_ptr) = (value == "true");
+                                valid = true;
+                            }
+                        }
+                        if (!valid) {
+                            throw std::exception();
+                        }
+                    } else {
+                        throw std::exception();
+                    }
+                }
+            } catch (std::exception) {
+                throw cxxopts::option_syntax_exception("Unable to parse supplied command line configuration options");
+            }
+        }
+    } catch (const cxxopts::OptionException& e) {
+        SPDLOG_ERROR("Error parsing options: {}", e.what());
+        std::cout << cmd_options.help() << std::endl;
+        exit(-1);
+    }
+}
+
+void validateNumericalOptions(AllOptInfo opt_info) {
+    for (OptInfo<int64_t> v : opt_info.i64_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
     }
+    for (OptInfo<int> v : opt_info.i_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
+    }
+    for (OptInfo<float> v : opt_info.f_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
+    }
+}
+
+MariusOptions parseConfig(int64_t argc, char *argv[]) {
+
+    // Associate each option with:
+    // variable, default value, section name, user option name, (optional) range
+
+    // List for each type of variable
+    std::vector<OptInfo<std::string>> s_var_map;
+    std::vector<OptInfo<int64_t>> i64_var_map;
+    std::vector<OptInfo<int>> i_var_map;
+    std::vector<OptInfo<float>> f_var_map;
+    std::vector<OptInfo<bool>> b_var_map;
+
+    float FLOAT_MAX = std::numeric_limits<float>::max();
 
     // General options
-    torch::DeviceType device;
+    torch::DeviceType device; // Device to use for training
     string s_device;
+    s_var_map.push_back((OptInfo<std::string>){&s_device, "CPU", "general", "device"});
+    std::vector<int> gpu_ids; // Ids of the gpus to use
     string s_gpu_ids;
-    std::vector<int> gpu_ids;
-    int64_t rand_seed;
-    int64_t num_train;
-    int64_t num_valid;
-    int64_t num_test;
-    int64_t num_nodes;
-    int64_t num_relations;
-    string experiment_name;
+    s_var_map.push_back((OptInfo<std::string>){&s_gpu_ids, "0", "general", "gpu_ids"});
```

Comment:
Nit: Change the declarations of the variables to be consectutive, rather than intermixed with the push_back functions calls. It's a little hard on the eyes.

E.g.
```
// General Options
int64_t random_seed;
int64_t num_train;
int64_t num_valid;
...

i64_var_map.push_back((OptInfo<int64_t>){&random_seed, time(0), "general", "random_seed", {0, INT64_MAX}});
i64_var_map.push_back((OptInfo<int64_t>){&num_train, -1, "general", "num_train", {0, INT64_MAX}});
i64_var_map.push_back((OptInfo<int64_t>){&num_valid, 0, "general", "num_valid", {0, INT64_MAX}});
...

// Model Options
float scale_factor;
...

f_var_map.push_back((OptInfo<float>){&scale_factor, .001, "model", "scale_factor", {0, FLOAT_MAX}});
...

```